### ISSUE TITLE
Incorrect type error assigning pd.read_csv to pandas DataFrame

### ISSUE THREAD

**Issue by betaprior** [microsoft/pyright#1195]
**Describe the bug**

Assigning result of `pd.read_csv()` to a variable declared as `pd.DataFrame` produces the following type error:

```
Expression of type "TextFileReader | Series | Any | Unknown | DataFrame | NDFrame" cannot be assigned to declared type "DataFrame"
   Type "TextFileReader | Series | Any | Unknown | DataFrame | NDFrame" cannot be assigned to type "DataFrame"
     "TextFileReader" is incompatible with "DataFrame"
     "Series" is incompatible with "DataFrame"
     "NDFrame" is incompatible with "DataFrame"
```

**Screenshots or Code**
The following triggers the problem for me:
```python
foo: pd.DataFrame = pd.read_csv('bar')
```

**VS Code extension or command-line**

This is under pyright 1.1.86 using lsp-mode in Emacs

**Comment by betaprior** [comment]
It appears that I was misunderstanding the behavior of the type system here (the variant type foo | bar I assume is an alias for Union[...] which needs an explicit cast when converting to a subtype)

**Comment by borisbrodski** [comment]
I'm getting the same error. Could you please explain a little bit more, why this behavior is correct
and how to avoid the error message?

And if I change my code like following, I still get the pyright-error:

```python
df = pd.read_csv('bar')
print(type(df))
c = df['column-name'] # "__getitem__" method not defined on type "TextFileReader" [Pyright: reportGeneralTypeIssues]
```
The code is actually working an printing `<class 'pandas.core.frame.DataFrame'>`

Thank you.

**Comment by jakebailey** [comment]
Without complete type info for pandas, it's unlikely that pyright will behave as you'd want when type checking code that uses it. Pylance ships hand-written stubs for pandas maintained here: https://github.com/microsoft/python-type-stubs

Which may help.
